app/app.go

```diff
@@ -1096,14 +1119,87 @@ func (app *App) InitChainer(ctx sdk.Context, req abci.RequestInitChain) abci.Res
 	return app.mm.InitGenesis(ctx, app.appCodec, genesisState, app.genesisImportConfig)
 }
 
-func (app *App) PrepareProposalHandler(_ sdk.Context, req *abci.RequestPrepareProposal) (*abci.ResponsePrepareProposal, error) {
+func (app *App) PrepareProposalHandler(ctx sdk.Context, req *abci.RequestPrepareProposal) (*abci.ResponsePrepareProposal, error) {
+	// Get all pending bundles for this height
+	ctx.Logger().Debug("Preparing proposal", "height", ctx.BlockHeight())
+
+	bundles := app.MevKeeper.PendingBundles(ctx.BlockHeight())
+
+	if len(bundles) > 0 {
+		// TODO remove this, just for debug
+		ctx.Logger().Error("found pending bundles from mevkeeper", "count", len(bundles))
+	}
+
+	maxTxBytes := req.MaxTxBytes
+	var selectedTxs [][]byte
+	var selectedTxsTotalSize = int64(0)
+	var remainingTxs [][]byte
+
+	// First, add any system transactions (governance, etc.)
+	for _, tx := range req.Txs {
+		if app.isSystemTx(tx) {
+			selectedTxs = append(selectedTxs, tx)
+			selectedTxsTotalSize += int64(len(tx))
+		} else {
+			remainingTxs = append(remainingTxs, tx)
+		}
+	}
+
+	// Next, add bundle transactions with the highest priority
+	for _, bundle := range bundles {
+		// Skip bundles not meant for this height
+		if bundle.BlockHeight != uint64(ctx.BlockHeight()) {
+			continue
+		}
+
+		// Calculate total size of this bundle
+		bundleSize := int64(0)
+		for _, txStr := range bundle.Transactions {
+			bundleSize += int64(len(txStr))
+		}
+
+		// Check if entire bundle fits
+		if selectedTxsTotalSize+bundleSize <= maxTxBytes {
+			// Add all transactions from this bundle
+			for _, tx := range bundle.Transactions {
+				//txBytes := []byte(txStr)
+				selectedTxs = append(selectedTxs, tx)
+			}
+			selectedTxsTotalSize += bundleSize
+		}
+	}
+
+	// Finally, add remaining transactions up to size limit
+	for _, tx := range remainingTxs {
+		if selectedTxsTotalSize+int64(len(tx)) <= maxTxBytes {
+			selectedTxs = append(selectedTxs, tx)
+			selectedTxsTotalSize += int64(len(tx))
+		}
+	}
+
+	// Convert selectedTxs to TxRecords
+	txRecords := make([]*abci.TxRecord, len(selectedTxs))
+	for i, tx := range selectedTxs {
+		txRecords[i] = &abci.TxRecord{
+			Action: abci.TxRecord_UNMODIFIED,
+			Tx:     tx,
+		}
+	}
+
+	//TODO purge used bundles, purge expired bundles
```

Comment:
is this blocking merging?

Reply:
We don't really do cleanup yet. But I'm wondering where exactly it should be in the codepath. Can block preparation be retried? Ideal place would be to purge all heights below a finalized block, regardless of who produced it, but that's probably for another PR